quickwit/quickwit-search/src/leaf.rs

```diff
@@ -459,46 +612,86 @@ pub async fn leaf_search(
             }.in_current_span())
         })
         .collect();
-    let split_search_results = futures::future::join_all(leaf_search_single_split_futures).await;
-
-    // the result wrapping is only for the collector api merge_fruits
-    // (Vec<tantivy::Result<LeafSearchResponse>>)
-    let (split_search_responses, errors): (
-        Vec<tantivy::Result<LeafSearchResponse>>,
-        Vec<(String, SearchError)>,
-    ) = split_search_results
-        .into_iter()
-        .partition_map(|split_search_res| match split_search_res {
-            Ok(Ok(split_search_resp)) => Either::Left(Ok(split_search_resp)),
-            Ok(Err(err)) => Either::Right(err),
-            Err(e) => {
-                warn!("A leaf_search_single_split panicked");
-                Either::Right(("unknown".to_string(), e.into()))
-            }
-        });
 
     // Creates a collector which merges responses into one
     let merge_collector =
         make_merge_collector(&request, &searcher_context.get_aggregation_limits())?;
 
-    // Merging is a cpu-bound task.
-    // It should be executed by Tokio's blocking threads.
-    let span = info_span!("merge_search_responses");
-    let mut merged_search_response = crate::run_cpu_intensive(move || {
-        let _span_guard = span.enter();
-        merge_collector.merge_fruits(split_search_responses)
-    })
-    .await
-    .context("Failed to merge split search responses.")??;
+    let wash_score = split_filter.wash_score();
+    if split_filter.iterative_merge() {
```

Comment:
I don't understand why we need a different logic for the iterative case and for the non iterative case.

Reply:
I expect iterative to be somewhat slower than non iterative merge, though this is probably a non-problem. Most of the filtering already happened before and this part is less performance critical. I'll bench things to see if having both code paths is really worth the trouble

Reply:
yes if you can check it is worth. I'd prefer have a single path here if possible.